utils/src/main/java/software/amazon/awssdk/utils/async/DelegatingBufferingSubscriber.java

```diff
@@ -0,0 +1,121 @@
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.utils.async;
+
+import static software.amazon.awssdk.utils.async.StoringSubscriber.EventType.ON_COMPLETE;
+import static software.amazon.awssdk.utils.async.StoringSubscriber.EventType.ON_NEXT;
+
+import java.nio.ByteBuffer;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicLong;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import software.amazon.awssdk.annotations.SdkProtectedApi;
+
+@SdkProtectedApi
+public class DelegatingBufferingSubscriber extends AbstractFlatteningSubscriber<ByteBuffer, ByteBuffer> {
+    /**
+     * The maximum amount of bytes allowed to be stored in the StoringSubscriber
+     */
+    private final long maximumBuffer;
+
+    /**
+     * Current amount of bytes buffered in the StoringSubscriber
+     */
+    private final AtomicLong currentlyBuffered = new AtomicLong(0);
+
+    /**
+     * Stores the bytes received from the upstream publisher, awaiting sending them to the delegate once the buffer size is
+     * reached.
+     */
+    private final StoringSubscriber<ByteBuffer> storage = new StoringSubscriber<>(Integer.MAX_VALUE);
+
+    public DelegatingBufferingSubscriber(long maximumBuffer, Subscriber<? super ByteBuffer> subscriber) {
+        super(subscriber);
+        this.maximumBuffer = maximumBuffer;
+    }
+
+    @Override
+    public void onNext(ByteBuffer item) {
+        if (currentlyBuffered.get() + item.remaining() > maximumBuffer) {
+            flushStorageToDelegate();
+        }
+        // calling super.onNext will eventually fulfill any remaining demand
+        super.onNext(item);
+    }
+
+    @Override
+    public void onSubscribe(Subscription subscription) {
+        storage.onSubscribe(new DemandIgnoringSubscription(subscription));
+        super.onSubscribe(subscription);
+    }
+
+    @Override
+    protected void doWithItem(ByteBuffer buffer) {
+        if (currentlyBuffered.get() + buffer.remaining() > maximumBuffer) {
+            flushStorageToDelegate();
+        }
+        storage.onNext(buffer.duplicate());
+        currentlyBuffered.addAndGet(buffer.remaining());
+    }
+
+    @Override
+    protected void fulfillDownstreamDemand() {
+        flushStorageToDelegate();
+    }
+
+    /**
+     * Returns true if we need to call onNext downstream.
+     */
+    @Override
+    protected boolean onNextNeeded() {
+        return super.onNextNeeded() && storage.peek().isPresent() && storage.peek().get().type() == ON_NEXT;
```

Comment:
Instead of doing `super.onNextNeeded()`, suggesting using template methods:

In the parent class:

```java
    private boolean onNextNeeded() {
        return downstreamDemand.get() > 0 && additionalOnNextNeededCheck();
    }

/**
  Can be overridden by subclasses to provide additional checks
*/
    protected boolean additionalOnNextNeededCheck() {
        return true;
    }
```